PSReadLine/Render.cs

```diff
@@ -340,135 +347,330 @@ void MaybeEmphasize(int i, string currColor)
+        {
+            int bufferWidth = _console.BufferWidth;
+            int bufferHeight = _console.BufferHeight;
+            int cursorX = _console.CursorLeft;
+            int cursorY = _console.CursorTop;
+
+            RenderedLineData[] previousRenderLines = _previousRender.lines;
+            int previousLogicalLine = 0;
+            int previousPhysicalLine = 0;
 
-                    // promptBufferCells is the number of visible characters in the prompt
-                    int promptBufferCells = LengthInBufferCells(promptText);
-                    bool renderErrorPrompt = false;
+            RenderedLineData[] renderLines = renderData.lines;
+            int logicalLine = 0;
+            int physicalLine = 0;
+            int pseudoPhysicalLineOffset = 0;
 
-                    if (_console.CursorLeft >= promptBufferCells)
+            bool hasToWriteAll = true;
+
+            if (cursorY > _initialY && renderLines.Length > 1)
+            {
+                // The current 'cursorTop' is below the initial 'cursorTop' and there are multiple logical lines.
+                // This indicates the user is editing in the middle or end of the existing text.
+                // In this case, it's possible that we can skip rendering until reaching the first changed logical line.
+
+                int minLineLength = previousRenderLines.Length;
+                int linesToCheck = -1;
+
+                if (renderLines.Length < previousRenderLines.Length)
+                {
+                    // Handle a special case:
+                    //   - top part of the text has been scrolled up-off the buffer;
+                    //   - the cursor is at the beginning of the first line in buffer;
+                    //   - the first line contains nothing but a new-line character;
+                    //   - the edit operation was backward delete a character.
+                    //
+                    // The editing was essentially removing the current empty line and move the cursor to the end of the previous line.
+                    // In this case, what we want is to start rendering from the previous logical line, where the cursor is supposed
+                    // to be moved to. However, if we only compare to find the first changed logical line, we will miss the right one
+                    // here, because that logical line is the same as before since the next logical line wrapped to it is empty.
+                    //
+                    // If the current logical lines are less than the previous, and the cursor is at the beginning of the first line in buffer,
+                    // then it's possible we are facing this special case and thus would need to do additional checks later.
+
+                    minLineLength = renderLines.Length;
+                    if (cursorX == Options.ContinuationPrompt.Length && cursorY == 0)
```

Comment:
With the `cursorY > _initialY` check above, this won't happen, because of the `cursorY == 0` check here implies `_initialY < 0`, meaning that this won't be the first physical line where the user prompt is at.
Now that the `cursorY > _initialY` check is removed, we need to check for `_initialY < 0` here.

Reply:
This piece of code has been changed, to cover the more general scenarios -- after editing, the cursor is supposed to be moved off the buffer.
Please take another look.